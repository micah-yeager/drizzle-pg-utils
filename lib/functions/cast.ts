import { type SQL, type SQLWrapper, sql } from "drizzle-orm"
import type {
	ColumnBuilderBase,
	ColumnBuilderBaseConfig,
	ColumnDataType,
	MakeColumnConfig,
} from "drizzle-orm/column-builder"
import {
	type AnyPgTable,
	PgColumn,
	type PgColumnBuilder,
	PgTable,
} from "drizzle-orm/pg-core"
import type { TypeFromSQL } from "../util"

/** Mimic the internal `build` method on PgColumnBuilder. */
interface PgColumnBuilder_ extends PgColumnBuilder {
	build(
		table: AnyPgTable<{ name: string }>,
	): PgColumn<
		MakeColumnConfig<ColumnBuilderBaseConfig<ColumnDataType, string>, string>
	>
}

/** A blank table to use for casting. */
const blankTable = new PgTable("", undefined, "")

/** @see https://www.postgresql.org/docs/17/sql-expressions.html#SQL-SYNTAX-TYPE-CASTS */
export function cast<T extends PgColumn | ColumnBuilderBase>(
	expression: SQLWrapper,
	type: T | ((...args: any[]) => T),
): SQL<TypeFromSQL<T, { nullable: false }>> {
	let sqlType: string
	// Column instance.
	if (type instanceof PgColumn) {
		sqlType = type.getSQLType()
	}
	// Column builder.
	else {
		const builder = (
			typeof type === "object" ? type : type()
		) as PgColumnBuilder_
		sqlType = builder.build(blankTable).getSQLType()
	}

	return sql`cast(${expression} as ${sql.raw(sqlType)})`
}
