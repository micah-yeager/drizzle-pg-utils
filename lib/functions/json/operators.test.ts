import { sql } from "drizzle-orm"
import { json } from "drizzle-orm/pg-core"
import { expect, it } from "vitest"
import { cast } from "../cast"
import {
	jsonElement,
	jsonElementAsText,
	jsonElementFromPathArray,
	jsonElementFromPathArrayAsText,
} from "./operators"

const testObj = { test: 0, nested: { key: "value" } } as const

it("should execute correctly with jsonElement", async () => {
	const toTest = jsonElement(
		cast(sql`${JSON.stringify(testObj)}`, json("").$type<typeof testObj>()),
		"nested",
	)

	const result = await globalThis.db.execute(sql`select ${toTest}`)

	expect(result.rows).toMatchSnapshot()
})

it("should execute correctly with jsonElementAsText", async () => {
	const toTest = jsonElementAsText(
		cast(sql`${JSON.stringify(testObj)}`, json("").$type<typeof testObj>()),
		"nested",
	)

	const result = await globalThis.db.execute(sql`select ${toTest}`)

	expect(result.rows).toMatchSnapshot()
})

it("should execute correctly with jsonElementFromPathArray", async () => {
	const toTest = jsonElementFromPathArray(
		cast(sql`${JSON.stringify(testObj)}`, json("").$type<typeof testObj>()),
		["nested", "key"],
	)

	const result = await globalThis.db.execute(sql`select ${toTest}`)

	expect(result.rows).toMatchSnapshot()
})

it("should execute correctly with jsonElementFromPathArrayAsText", async () => {
	const toTest = jsonElementFromPathArrayAsText(
		cast(sql`${JSON.stringify(testObj)}`, json("").$type<typeof testObj>()),
		["nested", "key"],
	)

	const result = await globalThis.db.execute(sql`select ${toTest}`)

	expect(result.rows).toMatchSnapshot()
})
