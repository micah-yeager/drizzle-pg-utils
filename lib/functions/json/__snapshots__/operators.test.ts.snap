// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should execute correctly with jsonElement 1`] = `
[
  {
    "?column?": {
      "key": "value",
    },
  },
]
`;

exports[`should execute correctly with jsonElementAsText 1`] = `
[
  {
    "?column?": {
      "key": "value",
    },
  },
]
`;

exports[`should execute correctly with jsonElementFromPathArray 1`] = `
[
  {
    "?column?": "value",
  },
]
`;

exports[`should execute correctly with jsonElementFromPathArrayAsText 1`] = `
[
  {
    "?column?": "value",
  },
]
`;
